src/projection.jl

```diff
@@ -31,114 +40,383 @@ function Base.show(io::IO, project::ProjectTo{T}) where {T}
+for T in (Float16, Float32, Float64, ComplexF16, ComplexF32, ComplexF64)
+    # Preserve low-precision floats as accidental promotion is a common perforance bug
+    @eval ProjectTo(::$T) = ProjectTo{$T}()
+end
+ProjectTo(x::Integer) = ProjectTo(float(x))
+ProjectTo(x::Complex{<:Integer}) = ProjectTo(float(x))
 (::ProjectTo{T})(dx::Number) where {T<:Number} = convert(T, dx)
 (::ProjectTo{T})(dx::Number) where {T<:Real} = convert(T, real(dx))
 
-# Arrays 
-ProjectTo(xs::T) where {T<:Array} = ProjectTo{T}(; elements=map(ProjectTo, xs))
-function (project::ProjectTo{T})(dx::Array) where {T<:Array}
-    _call(f, x) = f(x)
-    return T(map(_call, project.elements, dx))
+# Arrays
+# If we don't have a more specialized `ProjectTo` rule, we just assume that there is
+# no structure worth re-imposing. Then any array is acceptable as a gradient.
+
+# For arrays of numbers, just store one projector:
+function ProjectTo(x::AbstractArray{T}) where {T<:Number}
+    element = ProjectTo(zero(T))
+    if element isa ProjectTo{<:AbstractZero}
+        return ProjectTo{NoTangent}() # short-circuit if all elements project to zero
+    else
+        return ProjectTo{AbstractArray}(; element=element, axes=axes(x))
+    end
 end
-function (project::ProjectTo{T})(dx::AbstractZero) where {T<:Array}
-    return T(map(proj -> proj(dx), project.elements))
+
+# In other cases, store a projector per element:
+function ProjectTo(xs::AbstractArray)
+    elements = map(_always_projector, xs)
+    if elements isa AbstractArray{<:ProjectTo{<:AbstractZero}}
+        return ProjectTo{NoTangent}() # short-circuit if all elements project to zero
+    elseif elements isa AbstractArray{<:ProjectTo{Any}}
+        return ProjectTo{AbstractArray}(; element=ProjectTo(), axes=axes(xs)) # ... or none project
+    else
+        # Arrays of arrays come here, and will apply projectors individually:
+        return ProjectTo{AbstractArray}(; elements=elements, axes=axes(xs))
+    end
 end
-(project::ProjectTo{<:Array})(dx::AbstractArray) = project(collect(dx))
 
-# Arrays{<:Number}: optimized case so we don't need a projector per element
-function ProjectTo(x::T) where {E<:Number,T<:Array{E}}
-    return ProjectTo{T}(; element=ProjectTo(zero(E)), size=size(x))
+function (project::ProjectTo{AbstractArray})(dx::AbstractArray{S,M}) where {S,M}
+    # First deal with shape. The rule is that we reshape to add or remove trivial dimensions
+    # like dx = ones(4,1), where x = ones(4), but throw an error on dx = ones(1,4) etc.
+    dy = if axes(dx) == project.axes
+        dx
+    else
+        for d in 1:max(M, length(project.axes))
+            size(dx, d) == length(get(project.axes, d, 1)) || throw(_projection_mismatch(project.axes, size(dx)))
+        end
+        reshape(dx, project.axes)
+    end
+    # Then deal with the elements. One projector if AbstractArray{<:Number},
+    # or one per element for arrays of arrays:
+    dz = if hasfield(typeof(backing(project)), :element)
```

Comment:
```suggestion
    dz = if hasproperty(project), :element)
```

Reply:
Thoughts on whether this will infer equally well? Was trying to make it use `ProjectTo{AbstractArray, B}` but got awkward ambiguities.

Reply:
Worked OK, but failed on 1.0, so reverted.